File: packages/workz_sdk/android/src/main/kotlin/com/workz/workz_sdk/WorkzSdkPlugin.kt
// packages/workz_sdk/android/src/main/kotlin/com/workz/workz_sdk/WorkzSdkPlugin.kt

package com.workz.workz_sdk

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException

/** WorkzSdkPlugin */
class WorkzSdkPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private var workzSDK: WorkzSDKNative? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "workz_sdk")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "init" -> {
        val apiUrl = call.argument<String>("apiUrl")
        val token = call.argument<String>("token")
        
        if (apiUrl != null && token != null) {
          workzSDK = WorkzSDKNative(apiUrl, token)
          result.success(true)
        } else {
          result.error("INVALID_ARGUMENTS", "apiUrl and token are required", null)
        }
      }
      "getToken" -> {
        result.success(workzSDK?.getToken())
      }
      "getUser" -> {
        workzSDK?.getUser { user, error ->
          if (error != null) {
            result.error("GET_USER_ERROR", error, null)
          } else {
            result.success(user)
          }
        }
      }
      "isReady" -> {
        result.success(workzSDK?.isReady() ?: false)
      }
      "apiGet" -> {
        val path = call.argument<String>("path")
        if (path != null) {
          workzSDK?.apiGet(path) { response, error ->
            if (error != null) {
              result.error("API_GET_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "path is required", null)
        }
      }
      "apiPost" -> {
        val path = call.argument<String>("path")
        val body = call.argument<Map<String, Any>>("body")
        if (path != null) {
          workzSDK?.apiPost(path, body) { response, error ->
            if (error != null) {
              result.error("API_POST_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "path is required", null)
        }
      }
      "apiPut" -> {
        val path = call.argument<String>("path")
        val body = call.argument<Map<String, Any>>("body")
        if (path != null) {
          workzSDK?.apiPut(path, body) { response, error ->
            if (error != null) {
              result.error("API_PUT_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "path is required", null)
        }
      }
      "apiDelete" -> {
        val path = call.argument<String>("path")
        if (path != null) {
          workzSDK?.apiDelete(path) { response, error ->
            if (error != null) {
              result.error("API_DELETE_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "path is required", null)
        }
      }
      "kvSet" -> {
        val key = call.argument<String>("key")
        val value = call.argument<String>("value")
        val ttl = call.argument<Int>("ttl")
        if (key != null && value != null) {
          workzSDK?.kvSet(key, value, ttl) { response, error ->
            if (error != null) {
              result.error("KV_SET_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "key and value are required", null)
        }
      }
      "kvGet" -> {
        val key = call.argument<String>("key")
        if (key != null) {
          workzSDK?.kvGet(key) { response, error ->
            if (error != null) {
              result.error("KV_GET_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "key is required", null)
        }
      }
      "kvDelete" -> {
        val key = call.argument<String>("key")
        if (key != null) {
          workzSDK?.kvDelete(key) { response, error ->
            if (error != null) {
              result.error("KV_DELETE_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "key is required", null)
        }
      }
      "kvList" -> {
        workzSDK?.kvList { response, error ->
          if (error != null) {
            result.error("KV_LIST_ERROR", error, null)
          } else {
            result.success(response)
          }
        }
      }
      "docsSave" -> {
        val id = call.argument<String>("id")
        val document = call.argument<Map<String, Any>>("document")
        if (id != null && document != null) {
          workzSDK?.docsSave(id, document) { response, error ->
            if (error != null) {
              result.error("DOCS_SAVE_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "id and document are required", null)
        }
      }
      "docsGet" -> {
        val id = call.argument<String>("id")
        if (id != null) {
          workzSDK?.docsGet(id) { response, error ->
            if (error != null) {
              result.error("DOCS_GET_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "id is required", null)
        }
      }
      "docsDelete" -> {
        val id = call.argument<String>("id")
        if (id != null) {
          workzSDK?.docsDelete(id) { response, error ->
            if (error != null) {
              result.error("DOCS_DELETE_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "id is required", null)
        }
      }
      "docsList" -> {
        workzSDK?.docsList { response, error ->
          if (error != null) {
            result.error("DOCS_LIST_ERROR", error, null)
          } else {
            result.success(response)
          }
        }
      }
      "docsQuery" -> {
        val query = call.argument<Map<String, Any>>("query")
        if (query != null) {
          workzSDK?.docsQuery(query) { response, error ->
            if (error != null) {
              result.error("DOCS_QUERY_ERROR", error, null)
            } else {
              result.success(response)
            }
          }
        } else {
          result.error("INVALID_ARGUMENTS", "query is required", null)
        }
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

class WorkzSDKNative(private val apiUrl: String, private val token: String) {
  private val client = OkHttpClient()
  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  private var user: Map<String, Any>? = null
  private var isInitialized = false

  init {
    // Initialize and fetch user info
    scope.launch {
      try {
        fetchUserInfo()
        isInitialized = true
      } catch (e: Exception) {
        println("Failed to initialize WorkzSDK: ${e.message}")
      }
    }
  }

  fun getToken(): String = token

  fun isReady(): Boolean = isInitialized

  fun getUser(callback: (Map<String, Any>?, String?) -> Unit) {
    if (user != null) {
      callback(user, null)
    } else {
      scope.launch {
        try {
          fetchUserInfo()
          callback(user, null)
        } catch (e: Exception) {
          callback(null, e.message)
        }
      }
    }
  }

  private suspend fun fetchUserInfo() {
    val request = Request.Builder()
      .url("$apiUrl/me")
      .addHeader("Authorization", "Bearer $token")
      .build()

    val response = client.newCall(request).execute()
    if (response.isSuccessful) {
      val jsonString = response.body?.string()
      if (jsonString != null) {
        val jsonObject = JSONObject(jsonString)
        user = jsonObject.toMap()
      }
    }
  }

  fun apiGet(path: String, callback: (Map<String, Any>?, String?) -> Unit) {
    scope.launch {
      try {
        val request = Request.Builder()
          .url("$apiUrl$path")
          .addHeader("Authorization", "Bearer $token")
          .build()

        val response = client.newCall(request).execute()
        val jsonString = response.body?.string()
        if (response.isSuccessful && jsonString != null) {
          val jsonObject = JSONObject(jsonString)
          callback(jsonObject.toMap(), null)
        } else {
          callback(null, "HTTP ${response.code}: ${response.message}")
        }
      } catch (e: Exception) {
        callback(null, e.message)
      }
    }
  }

  fun apiPost(path: String, body: Map<String, Any>?, callback: (Map<String, Any>?, String?) -> Unit) {
    scope.launch {
      try {
        val jsonBody = if (body != null) {
          JSONObject(body).toString().toRequestBody("application/json".toMediaType())
        } else {
          "{}".toRequestBody("application/json".toMediaType())
        }

        val request = Request.Builder()
          .url("$apiUrl$path")
          .addHeader("Authorization", "Bearer $token")
          .post(jsonBody)
          .build()

        val response = client.newCall(request).execute()
        val jsonString = response.body?.string()
        if (response.isSuccessful && jsonString != null) {
          val jsonObject = JSONObject(jsonString)
          callback(jsonObject.toMap(), null)
        } else {
          callback(null, "HTTP ${response.code}: ${response.message}")
        }
      } catch (e: Exception) {
        callback(null, e.message)
      }
    }
  }

  fun apiPut(path: String, body: Map<String, Any>?, callback: (Map<String, Any>?, String?) -> Unit) {
    scope.launch {
      try {
        val jsonBody = if (body != null) {
          JSONObject(body).toString().toRequestBody("application/json".toMediaType())
        } else {
          "{}".toRequestBody("application/json".toMediaType())
        }

        val request = Request.Builder()
          .url("$apiUrl$path")
          .addHeader("Authorization", "Bearer $token")
          .put(jsonBody)
          .build()

        val response = client.newCall(request).execute()
        val jsonString = response.body?.string()
        if (response.isSuccessful && jsonString != null) {
          val jsonObject = JSONObject(jsonString)
          callback(jsonObject.toMap(), null)
        } else {
          callback(null, "HTTP ${response.code}: ${response.message}")
        }
      } catch (e: Exception) {
        callback(null, e.message)
      }
    }
  }

  fun apiDelete(path: String, callback: (Map<String, Any>?, String?) -> Unit) {
    scope.launch {
      try {
        val request = Request.Builder()
          .url("$apiUrl$path")
          .addHeader("Authorization", "Bearer $token")
          .delete()
          .build()

        val response = client.newCall(request).execute()
        val jsonString = response.body?.string()
        if (response.isSuccessful && jsonString != null) {
          val jsonObject = JSONObject(jsonString)
          callback(jsonObject.toMap(), null)
        } else {
          callback(null, "HTTP ${response.code}: ${response.message}")
        }
      } catch (e: Exception) {
        callback(null, e.message)
      }
    }
  }

  // KV Storage methods
  fun kvSet(key: String, value: String, ttl: Int?, callback: (Map<String, Any>?, String?) -> Unit) {
    val body = mutableMapOf<String, Any>(
      "key" to key,
      "value" to value,
      "scopeType" to "user",
      "scopeId" to (user?.get("id") ?: 0)
    )
    if (ttl != null) {
      body["ttl"] = ttl
    }
    apiPost("/appdata/kv", body, callback)
  }

  fun kvGet(key: String, callback: (Map<String, Any>?, String?) -> Unit) {
    val userId = user?.get("id") ?: 0
    apiGet("/appdata/kv?key=$key&scopeType=user&scopeId=$userId", callback)
  }

  fun kvDelete(key: String, callback: (Map<String, Any>?, String?) -> Unit) {
    val userId = user?.get("id") ?: 0
    apiDelete("/appdata/kv?key=$key&scopeType=user&scopeId=$userId", callback)
  }

  fun kvList(callback: (Map<String, Any>?, String?) -> Unit) {
    val userId = user?.get("id") ?: 0
    apiGet("/appdata/kv?scopeType=user&scopeId=$userId", callback)
  }

  // Document Storage methods
  fun docsSave(id: String, document: Map<String, Any>, callback: (Map<String, Any>?, String?) -> Unit) {
    val body = mapOf(
      "docType" to "user_data",
      "docId" to id,
      "document" to document,
      "scopeType" to "user",
      "scopeId" to (user?.get("id") ?: 0)
    )
    apiPost("/appdata/docs/upsert", body, callback)
  }

  fun docsGet(id: String, callback: (Map<String, Any>?, String?) -> Unit) {
    val body = mapOf(
      "docType" to "user_data",
      "filters" to mapOf("docId" to id),
      "scopeType" to "user",
      "scopeId" to (user?.get("id") ?: 0)
    )
    apiPost("/appdata/docs/query", body, callback)
  }

  fun docsDelete(id: String, callback: (Map<String, Any>?, String?) -> Unit) {
    val userId = user?.get("id") ?: 0
    apiDelete("/appdata/docs/user_data/$id?scopeType=user&scopeId=$userId", callback)
  }

  fun docsList(callback: (Map<String, Any>?, String?) -> Unit) {
    val body = mapOf(
      "docType" to "user_data",
      "filters" to emptyMap<String, Any>(),
      "scopeType" to "user",
      "scopeId" to (user?.get("id") ?: 0)
    )
    apiPost("/appdata/docs/query", body, callback)
  }

  fun docsQuery(query: Map<String, Any>, callback: (Map<String, Any>?, String?) -> Unit) {
    val body = mapOf(
      "docType" to "user_data",
      "filters" to query,
      "scopeType" to "user",
      "scopeId" to (user?.get("id") ?: 0)
    )
    apiPost("/appdata/docs/query", body, callback)
  }
}

// Extension function to convert JSONObject to Map
fun JSONObject.toMap(): Map<String, Any> {
  val map = mutableMapOf<String, Any>()
  val keys = this.keys()
  while (keys.hasNext()) {
    val key = keys.next()
    val value = this.get(key)
    map[key] = when (value) {
      is JSONObject -> value.toMap()
      else -> value
    }
  }
  return map
}

File: packages/workz_sdk/ios/Classes/WorkzSdkPlugin.swift
// packages/workz_sdk/ios/Classes/WorkzSdkPlugin.swift

import Flutter
import UIKit
import Foundation

public class WorkzSdkPlugin: NSObject, FlutterPlugin {
  private var workzSDK: WorkzSDKNative?
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "workz_sdk", binaryMessenger: registrar.messenger())
    let instance = WorkzSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
      guard let args = call.arguments as? [String: Any],
            let apiUrl = args["apiUrl"] as? String,
            let token = args["token"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "apiUrl and token are required", details: nil))
        return
      }
      
      workzSDK = WorkzSDKNative(apiUrl: apiUrl, token: token)
      result(true)
      
    case "getToken":
      result(workzSDK?.getToken())
      
    case "getUser":
      workzSDK?.getUser { user, error in
        if let error = error {
          result(FlutterError(code: "GET_USER_ERROR", message: error, details: nil))
        } else {
          result(user)
        }
      }
      
    case "isReady":
      result(workzSDK?.isReady() ?? false)
      
    case "apiGet":
      guard let args = call.arguments as? [String: Any],
            let path = args["path"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "path is required", details: nil))
        return
      }
      
      workzSDK?.apiGet(path: path) { response, error in
        if let error = error {
          result(FlutterError(code: "API_GET_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "apiPost":
      guard let args = call.arguments as? [String: Any],
            let path = args["path"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "path is required", details: nil))
        return
      }
      
      let body = args["body"] as? [String: Any]
      workzSDK?.apiPost(path: path, body: body) { response, error in
        if let error = error {
          result(FlutterError(code: "API_POST_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "apiPut":
      guard let args = call.arguments as? [String: Any],
            let path = args["path"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "path is required", details: nil))
        return
      }
      
      let body = args["body"] as? [String: Any]
      workzSDK?.apiPut(path: path, body: body) { response, error in
        if let error = error {
          result(FlutterError(code: "API_PUT_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "apiDelete":
      guard let args = call.arguments as? [String: Any],
            let path = args["path"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "path is required", details: nil))
        return
      }
      
      workzSDK?.apiDelete(path: path) { response, error in
        if let error = error {
          result(FlutterError(code: "API_DELETE_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "kvSet":
      guard let args = call.arguments as? [String: Any],
            let key = args["key"] as? String,
            let value = args["value"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "key and value are required", details: nil))
        return
      }
      
      let ttl = args["ttl"] as? Int
      workzSDK?.kvSet(key: key, value: value, ttl: ttl) { response, error in
        if let error = error {
          result(FlutterError(code: "KV_SET_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "kvGet":
      guard let args = call.arguments as? [String: Any],
            let key = args["key"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "key is required", details: nil))
        return
      }
      
      workzSDK?.kvGet(key: key) { response, error in
        if let error = error {
          result(FlutterError(code: "KV_GET_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "kvDelete":
      guard let args = call.arguments as? [String: Any],
            let key = args["key"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "key is required", details: nil))
        return
      }
      
      workzSDK?.kvDelete(key: key) { response, error in
        if let error = error {
          result(FlutterError(code: "KV_DELETE_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "kvList":
      workzSDK?.kvList { response, error in
        if let error = error {
          result(FlutterError(code: "KV_LIST_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "docsSave":
      guard let args = call.arguments as? [String: Any],
            let id = args["id"] as? String,
            let document = args["document"] as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "id and document are required", details: nil))
        return
      }
      
      workzSDK?.docsSave(id: id, document: document) { response, error in
        if let error = error {
          result(FlutterError(code: "DOCS_SAVE_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "docsGet":
      guard let args = call.arguments as? [String: Any],
            let id = args["id"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "id is required", details: nil))
        return
      }
      
      workzSDK?.docsGet(id: id) { response, error in
        if let error = error {
          result(FlutterError(code: "DOCS_GET_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "docsDelete":
      guard let args = call.arguments as? [String: Any],
            let id = args["id"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "id is required", details: nil))
        return
      }
      
      workzSDK?.docsDelete(id: id) { response, error in
        if let error = error {
          result(FlutterError(code: "DOCS_DELETE_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "docsList":
      workzSDK?.docsList { response, error in
        if let error = error {
          result(FlutterError(code: "DOCS_LIST_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    case "docsQuery":
      guard let args = call.arguments as? [String: Any],
            let query = args["query"] as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "query is required", details: nil))
        return
      }
      
      workzSDK?.docsQuery(query: query) { response, error in
        if let error = error {
          result(FlutterError(code: "DOCS_QUERY_ERROR", message: error, details: nil))
        } else {
          result(response)
        }
      }
      
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

class WorkzSDKNative {
  private let apiUrl: String
  private let token: String
  private var user: [String: Any]?
  private var isInitialized = false
  
  init(apiUrl: String, token: String) {
    self.apiUrl = apiUrl
    self.token = token
    
    // Initialize and fetch user info
    fetchUserInfo { [weak self] _, _ in
      self?.isInitialized = true
    }
  }
  
  func getToken() -> String {
    return token
  }
  
  func isReady() -> Bool {
    return isInitialized
  }
  
  func getUser(completion: @escaping ([String: Any]?, String?) -> Void) {
    if let user = user {
      completion(user, nil)
    } else {
      fetchUserInfo(completion: completion)
    }
  }
  
  private func fetchUserInfo(completion: @escaping ([String: Any]?, String?) -> Void) {
    apiGet(path: "/me", completion: { [weak self] response, error in
      if let response = response {
        self?.user = response
        completion(response, nil)
      } else {
        completion(nil, error)
      }
    })
  }
  
  func apiGet(path: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    guard let url = URL(string: "\(apiUrl)\(path)") else {
      completion(nil, "Invalid URL")
      return
    }
    
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    
    URLSession.shared.dataTask(with: request) { data, response, error in
      DispatchQueue.main.async {
        if let error = error {
          completion(nil, error.localizedDescription)
          return
        }
        
        guard let data = data else {
          completion(nil, "No data received")
          return
        }
        
        do {
          if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
            completion(json, nil)
          } else {
            completion(nil, "Invalid JSON response")
          }
        } catch {
          completion(nil, "JSON parsing error: \(error.localizedDescription)")
        }
      }
    }.resume()
  }
  
  func apiPost(path: String, body: [String: Any]?, completion: @escaping ([String: Any]?, String?) -> Void) {
    makeRequest(method: "POST", path: path, body: body, completion: completion)
  }
  
  func apiPut(path: String, body: [String: Any]?, completion: @escaping ([String: Any]?, String?) -> Void) {
    makeRequest(method: "PUT", path: path, body: body, completion: completion)
  }
  
  func apiDelete(path: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    makeRequest(method: "DELETE", path: path, body: nil, completion: completion)
  }
  
  private func makeRequest(method: String, path: String, body: [String: Any]?, completion: @escaping ([String: Any]?, String?) -> Void) {
    guard let url = URL(string: "\(apiUrl)\(path)") else {
      completion(nil, "Invalid URL")
      return
    }
    
    var request = URLRequest(url: url)
    request.httpMethod = method
    request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    
    if let body = body {
      do {
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
      } catch {
        completion(nil, "JSON encoding error: \(error.localizedDescription)")
        return
      }
    }
    
    URLSession.shared.dataTask(with: request) { data, response, error in
      DispatchQueue.main.async {
        if let error = error {
          completion(nil, error.localizedDescription)
          return
        }
        
        guard let data = data else {
          completion(nil, "No data received")
          return
        }
        
        do {
          if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
            completion(json, nil)
          } else {
            completion(nil, "Invalid JSON response")
          }
        } catch {
          completion(nil, "JSON parsing error: \(error.localizedDescription)")
        }
      }
    }.resume()
  }
  
  // KV Storage methods
  func kvSet(key: String, value: String, ttl: Int?, completion: @escaping ([String: Any]?, String?) -> Void) {
    var body: [String: Any] = [
      "key": key,
      "value": value,
      "scopeType": "user",
      "scopeId": user?["id"] ?? 0
    ]
    if let ttl = ttl {
      body["ttl"] = ttl
    }
    apiPost(path: "/appdata/kv", body: body, completion: completion)
  }
  
  func kvGet(key: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    let userId = user?["id"] ?? 0
    apiGet(path: "/appdata/kv?key=\(key)&scopeType=user&scopeId=\(userId)", completion: completion)
  }
  
  func kvDelete(key: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    let userId = user?["id"] ?? 0
    apiDelete(path: "/appdata/kv?key=\(key)&scopeType=user&scopeId=\(userId)", completion: completion)
  }
  
  func kvList(completion: @escaping ([String: Any]?, String?) -> Void) {
    let userId = user?["id"] ?? 0
    apiGet(path: "/appdata/kv?scopeType=user&scopeId=\(userId)", completion: completion)
  }
  
  // Document Storage methods
  func docsSave(id: String, document: [String: Any], completion: @escaping ([String: Any]?, String?) -> Void) {
    let body: [String: Any] = [
      "docType": "user_data",
      "docId": id,
      "document": document,
      "scopeType": "user",
      "scopeId": user?["id"] ?? 0
    ]
    apiPost(path: "/appdata/docs/upsert", body: body, completion: completion)
  }
  
  func docsGet(id: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    let body: [String: Any] = [
      "docType": "user_data",
      "filters": ["docId": id],
      "scopeType": "user",
      "scopeId": user?["id"] ?? 0
    ]
    apiPost(path: "/appdata/docs/query", body: body, completion: completion)
  }
  
  func docsDelete(id: String, completion: @escaping ([String: Any]?, String?) -> Void) {
    let userId = user?["id"] ?? 0
    apiDelete(path: "/appdata/docs/user_data/\(id)?scopeType=user&scopeId=\(userId)", completion: completion)
  }
  
  func docsList(completion: @escaping ([String: Any]?, String?) -> Void) {
    let body: [String: Any] = [
      "docType": "user_data",
      "filters": [:],
      "scopeType": "user",
      "scopeId": user?["id"] ?? 0
    ]
    apiPost(path: "/appdata/docs/query", body: body, completion: completion)
  }
  
  func docsQuery(query: [String: Any], completion: @escaping ([String: Any]?, String?) -> Void) {
    let body: [String: Any] = [
      "docType": "user_data",
      "filters": query,
      "scopeType": "user",
      "scopeId": user?["id"] ?? 0
    ]
    apiPost(path: "/appdata/docs/query", body: body, completion: completion)
  }
}